source/builtin/src/lib.rs

```diff
@@ -414,3 +382,298 @@ pub struct SyncSendIfSend<T> {
+// Marker for integer types (i8 ... u128, isize, usize, nat, int)
+// so that we get reasonable type error messages when someone uses a non-Integer type
+// in an arithmetic operation.
+pub trait Integer {}
+impl Integer for u8 {}
+impl Integer for u16 {}
+impl Integer for u32 {}
+impl Integer for u64 {}
+impl Integer for u128 {}
+impl Integer for usize {}
+impl Integer for i8 {}
+impl Integer for i16 {}
+impl Integer for i32 {}
+impl Integer for i64 {}
+impl Integer for i128 {}
+impl Integer for isize {}
+impl Integer for int {}
+impl Integer for nat {}
+
+// spec literals of the form "33", which could have any Integer type
+#[allow(non_camel_case_types)]
+#[spec]
+pub fn spec_literal_integer<
+    hint_please_add_suffix_on_literal_like_100u32_or_100int_or_100nat: Integer,
+>(
+    _s: &str,
+) -> hint_please_add_suffix_on_literal_like_100u32_or_100int_or_100nat {
+    unimplemented!()
+}
+
+// spec literals of the form "33int",
+// or spec literals in positions syntactically expected to be int (e.g. in "x + 33")
+#[spec]
+pub fn spec_literal_int(_s: &str) -> int {
+    unimplemented!()
+}
+
+// spec literals of the form "33nat"
+#[spec]
+pub fn spec_literal_nat(_s: &str) -> nat {
+    unimplemented!()
+}
+
+// Fixed-width add
+#[spec]
+pub fn add<IntegerType: Integer>(_left: IntegerType, _right: IntegerType) -> IntegerType {
+    unimplemented!()
+}
+
+// Fixed-width sub
+#[spec]
+pub fn sub<IntegerType: Integer>(_left: IntegerType, _right: IntegerType) -> IntegerType {
+    unimplemented!()
+}
+
+// Fixed-width mul
+#[spec]
+pub fn mul<IntegerType: Integer>(_left: IntegerType, _right: IntegerType) -> IntegerType {
+    unimplemented!()
+}
+
+// represent "expr as typ", including converting to and from int and nat
+#[spec]
+pub fn spec_cast_integer<From: Integer, To: Integer>(_from: From) -> To {
+    unimplemented!()
+}
+
+/*
+pub trait SpecEq<Rhs = Self> {
+    #[spec]
+    fn spec_eq(self, other: Rhs) -> bool;
+    #[spec]
+    fn spec_ne(self, other: Rhs) -> bool;
+}
+*/
+
+#[spec]
+pub fn spec_eq<Lhs: Structural, Rhs: Structural>(_lhs: Lhs, _rhs: Rhs) -> bool {
+    unimplemented!()
+}
+
+pub trait SpecOrd<Rhs = Self> {
+    #[spec]
+    fn spec_lt(self, rhs: Rhs) -> bool;
+    #[spec]
+    fn spec_le(self, rhs: Rhs) -> bool;
+    #[spec]
+    fn spec_gt(self, rhs: Rhs) -> bool;
+    #[spec]
+    fn spec_ge(self, rhs: Rhs) -> bool;
+}
+
+pub trait SpecIndex<Idx> {
+    type Output;
+    #[spec]
+    fn index(self, index: Idx) -> Self::Output;
+}
+
+pub trait SpecNeg {
+    type Output;
+    #[spec]
+    fn spec_neg(self) -> Self::Output;
+}
+
+pub trait SpecAdd<Rhs = Self> {
+    type Output;
+    #[spec]
+    fn spec_add(self, rhs: Rhs) -> Self::Output;
+}
+
+pub trait SpecSub<Rhs = Self> {
+    type Output;
+    #[spec]
+    fn spec_sub(self, rhs: Rhs) -> Self::Output;
+}
+
+pub trait SpecMul<Rhs = Self> {
+    type Output;
+    #[spec]
+    fn spec_mul(self, rhs: Rhs) -> Self::Output;
+}
+
+/*
+pub trait SpecEuclideanDiv<Rhs = Self> {
```

Comment:
Yes, good question.  Basically, I don't think we need traits for most of the operators (/, %, &, ^, <<, >>, etc.) for the purpose of turning on `--no-enhanced-typecheck`, because I don't think they would actually improve the type checking for integers, because having them automatically widen their integer arguments to type `int` seems to hurt more than help in our existing code.  (Whereas widening for ==, !=, <=, >=, <, >, +, -, * seems to work well in removing the need to manually write `as int`.)

It's possible, though, that in the future, we might want to allow programmers to overload these operators for their own datatypes in ghost code, even if the operators don't do anything special for integers.  But I think this can wait.

Anyway, I just went ahead and deleted the commented-out code.

Reply:
For the bitwise operators it makes sense. I don't get the reasoning for / and % though. As a user my default expectation would be that / and % would have the same behavior as *. (I guess it doesn't matter much for %, which can't overflow, but / can overflow.)

Reply:
I revised this a little.  Now there is a trait for / and a trait for %.  The implementation of the % trait for integers has the form `(Self, Self) -> Self`, basically like Rust's normal %.  The implementation of the / trait for unsigned integers has the form `(Self, Self) -> Self`, because these can't overflow, but has the form `(Self, Self) -> int` for signed integers because these can overflow.  I also updated + and * for `nat` so that, for example, `n + 1` has type `nat` if `n` has type `nat` (because writing `(n + 1) as nat` is kind of annoying). The general rule across +, -, *, /, % is that we return `int` when the input types allow a possibility of overflow, but otherwise try to make the output type the same as the input types.